android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt

```diff
@@ -308,4 +328,358 @@ constructor(
+              resourceConfig.resource,
+              ReferenceClientParam(resourceConfig.searchParameter)
+            )
+          } else {
+            include(ReferenceClientParam(resourceConfig.searchParameter), resourceConfig.resource)
+          }
+        }
+      }
+
+      searchRelatedResources(
+        isRevInclude = isRevInclude,
+        search = search,
+        relatedResourcesConfigsMap = relatedResourcesConfigsMap,
+        relatedResourceWrapper = relatedResourceWrapper,
+        configComputedRuleValues = configComputedRuleValues
+      )
+    }
+  }
+
+  private suspend fun searchRelatedResources(
+    isRevInclude: Boolean,
+    search: Search,
+    relatedResourcesConfigsMap: Map<ResourceType, List<ResourceConfig>>,
+    relatedResourceWrapper: RelatedResourceWrapper,
+    configComputedRuleValues: Map<String, Any>
+  ) {
+    kotlin
+      .runCatching { fhirEngine.searchWithRevInclude<Resource>(isRevInclude, search) }
+      .onSuccess { searchResult ->
+        searchResult.values.forEach { theRelatedResourcesMap: Map<ResourceType, List<Resource>> ->
+          theRelatedResourcesMap.forEach { entry ->
+            val currentResourceConfigs = relatedResourcesConfigsMap[entry.key]
+
+            val key = // Use configured id as key otherwise default to ResourceType
+              if (relatedResourcesConfigsMap.containsKey(entry.key)) {
+                currentResourceConfigs?.firstOrNull()?.id ?: entry.key.name
+              } else entry.key.name
+
+            // All nested resources flattened to one map by adding to existing list
+            relatedResourceWrapper.relatedResourceMap[key] =
+              relatedResourceWrapper
+                .relatedResourceMap
+                .getOrPut(key) { LinkedList() }
+                .plus(entry.value)
+
+            currentResourceConfigs?.forEach { resourceConfig ->
+              if (resourceConfig.relatedResources.isNotEmpty())
+                retrieveRelatedResources(
+                  resources = entry.value,
+                  relatedResourcesConfigs = resourceConfig.relatedResources,
+                  relatedResourceWrapper = relatedResourceWrapper,
+                  configComputedRuleValues = configComputedRuleValues
+                )
+            }
+          }
+        }
+      }
+      .onFailure {
+        Timber.e(it, "Error fetching configured related resources: $relatedResourcesConfigsMap")
+      }
+  }
+
+  private fun List<ResourceConfig>.revIncludeRelatedResourceConfigs(isRevInclude: Boolean) =
+    if (isRevInclude) this.filter { it.isRevInclude && !it.resultAsCount }
+    else this.filter { !it.isRevInclude && !it.resultAsCount }
+
+  suspend fun updateResourcesRecursively(resourceConfig: ResourceConfig, subject: Resource) {
+    val resourceData =
+      resourceDataRulesExecutor.processResourceData(
+        repositoryResourceData = RepositoryResourceData(resource = subject),
+        ruleConfigs = resourceConfig.configRules!!,
```

Comment:
Potential `NullPointerException`, can we provide a default empty list and use a safe null check instead?